File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Symbols/SynthesizedPrivateImplementationDetailsSharedConstructor.cs
using System.Collections.Generic;
using System.Collections.Immutable;
using Microsoft.CodeAnalysis.CodeGen;
using Microsoft.CodeAnalysis.PooledObjects;

namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
{
	internal sealed class SynthesizedPrivateImplementationDetailsSharedConstructor : SynthesizedGlobalMethodBase
	{
		private readonly SourceModuleSymbol _containingModule;

		private readonly PrivateImplementationDetails _privateImplementationType;

		private readonly TypeSymbol _voidType;

		public override bool IsSub => true;

		public override bool ReturnsByRef => false;

		public override TypeSymbol ReturnType => _voidType;

		internal override bool HasSpecialName => true;

		public override MethodKind MethodKind => MethodKind.StaticConstructor;

		public override ModuleSymbol ContainingModule => _containingModule;

		internal SynthesizedPrivateImplementationDetailsSharedConstructor(SourceModuleSymbol containingModule, PrivateImplementationDetails privateImplementationType, NamedTypeSymbol voidType)
			: base(containingModule, ".cctor", privateImplementationType)
		{
			_containingModule = containingModule;
			_privateImplementationType = privateImplementationType;
			_voidType = voidType;
		}

		internal override BoundBlock GetBoundMethodBody(TypeCompilationState compilationState, BindingDiagnosticBag diagnostics, ref Binder methodBodyBinder = null)
		{
			methodBodyBinder = null;
			SyntheticBoundNodeFactory syntheticBoundNodeFactory = new SyntheticBoundNodeFactory(this, this, VisualBasicSyntaxTree.Dummy.GetRoot(), compilationState, diagnostics);
			ArrayBuilder<BoundStatement> instance = ArrayBuilder<BoundStatement>.GetInstance();
			foreach (KeyValuePair<int, InstrumentationPayloadRootField> instrumentationPayloadRoot in _privateImplementationType.GetInstrumentationPayloadRoots())
			{
				int key = instrumentationPayloadRoot.Key;
				ArrayTypeSymbol arrayTypeSymbol = (ArrayTypeSymbol)instrumentationPayloadRoot.Value.Type.GetInternalSymbol();
				instance.Add(syntheticBoundNodeFactory.Assignment(syntheticBoundNodeFactory.InstrumentationPayloadRoot(key, arrayTypeSymbol, isLValue: true), syntheticBoundNodeFactory.Array(arrayTypeSymbol.ElementType, ImmutableArray.Create(syntheticBoundNodeFactory.MaximumMethodDefIndex()), ImmutableArray<BoundExpression>.Empty)));
			}
			MethodSymbol methodSymbol = syntheticBoundNodeFactory.WellKnownMember<MethodSymbol>(WellKnownMember.System_Guid__ctor);
			if ((object)methodSymbol != null)
			{
				instance.Add(syntheticBoundNodeFactory.Assignment(syntheticBoundNodeFactory.ModuleVersionId(isLValue: true), syntheticBoundNodeFactory.New(methodSymbol, syntheticBoundNodeFactory.ModuleVersionIdString())));
			}
			instance.Add(syntheticBoundNodeFactory.Return());
			return syntheticBoundNodeFactory.Block(instance.ToImmutableAndFree());
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Symbols/SynthesizedPrivateImplementationDetailsSharedConstructor.vb
Imports Microsoft.Cci
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.CodeGen
Imports Microsoft.CodeAnalysis.PooledObjects
Imports Microsoft.CodeAnalysis.VisualBasic
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Collections.Immutable
Imports System.Threading

Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
	Friend NotInheritable Class SynthesizedPrivateImplementationDetailsSharedConstructor
		Inherits SynthesizedGlobalMethodBase
		Private ReadOnly _containingModule As SourceModuleSymbol

		Private ReadOnly _privateImplementationType As PrivateImplementationDetails

		Private ReadOnly _voidType As TypeSymbol

		Public Overrides ReadOnly Property ContainingModule As ModuleSymbol
			Get
				Return Me._containingModule
			End Get
		End Property

		Friend Overrides ReadOnly Property HasSpecialName As Boolean
			Get
				Return True
			End Get
		End Property

		Public Overrides ReadOnly Property IsSub As Boolean
			Get
				Return True
			End Get
		End Property

		Public Overrides ReadOnly Property MethodKind As Microsoft.CodeAnalysis.MethodKind
			Get
				Return Microsoft.CodeAnalysis.MethodKind.StaticConstructor
			End Get
		End Property

		Public Overrides ReadOnly Property ReturnsByRef As Boolean
			Get
				Return False
			End Get
		End Property

		Public Overrides ReadOnly Property ReturnType As TypeSymbol
			Get
				Return Me._voidType
			End Get
		End Property

		Friend Sub New(ByVal containingModule As SourceModuleSymbol, ByVal privateImplementationType As PrivateImplementationDetails, ByVal voidType As NamedTypeSymbol)
			MyBase.New(containingModule, ".cctor", privateImplementationType)
			Me._containingModule = containingModule
			Me._privateImplementationType = privateImplementationType
			Me._voidType = voidType
		End Sub

		Friend Overrides Function GetBoundMethodBody(ByVal compilationState As TypeCompilationState, ByVal diagnostics As Microsoft.CodeAnalysis.VisualBasic.BindingDiagnosticBag, Optional ByRef methodBodyBinder As Binder = Nothing) As BoundBlock
			Dim enumerator As IEnumerator(Of KeyValuePair(Of Integer, InstrumentationPayloadRootField)) = Nothing
			methodBodyBinder = Nothing
			Dim dummy As VisualBasicSyntaxTree = VisualBasicSyntaxTree.Dummy
			Dim cancellationToken As System.Threading.CancellationToken = New System.Threading.CancellationToken()
			Dim syntheticBoundNodeFactory As Microsoft.CodeAnalysis.VisualBasic.SyntheticBoundNodeFactory = New Microsoft.CodeAnalysis.VisualBasic.SyntheticBoundNodeFactory(Me, Me, dummy.GetRoot(cancellationToken), compilationState, diagnostics)
			Using instance As ArrayBuilder(Of BoundStatement) = ArrayBuilder(Of BoundStatement).GetInstance()
				enumerator = Me._privateImplementationType.GetInstrumentationPayloadRoots().GetEnumerator()
				While enumerator.MoveNext()
					Dim current As KeyValuePair(Of Integer, InstrumentationPayloadRootField) = enumerator.Current
					Dim key As Integer = current.Key
					Dim internalSymbol As ArrayTypeSymbol = DirectCast(current.Value.Type.GetInternalSymbol(), ArrayTypeSymbol)
					instance.Add(syntheticBoundNodeFactory.Assignment(syntheticBoundNodeFactory.InstrumentationPayloadRoot(key, internalSymbol, True), syntheticBoundNodeFactory.Array(internalSymbol.ElementType, ImmutableArray.Create(Of BoundExpression)(syntheticBoundNodeFactory.MaximumMethodDefIndex()), ImmutableArray(Of BoundExpression).Empty)))
				End While
			End Using
			Dim methodSymbol As Microsoft.CodeAnalysis.VisualBasic.Symbols.MethodSymbol = syntheticBoundNodeFactory.WellKnownMember(Of Microsoft.CodeAnalysis.VisualBasic.Symbols.MethodSymbol)(WellKnownMember.System_Guid__ctor, False)
			If (methodSymbol IsNot Nothing) Then
				instance.Add(syntheticBoundNodeFactory.Assignment(syntheticBoundNodeFactory.ModuleVersionId(True), syntheticBoundNodeFactory.[New](methodSymbol, New BoundExpression() { syntheticBoundNodeFactory.ModuleVersionIdString() })))
			End If
			instance.Add(syntheticBoundNodeFactory.[Return](Nothing))
			Return syntheticBoundNodeFactory.Block(instance.ToImmutableAndFree())
		End Function
	End Class
End Namespace